lang/LangPrimSource/SC_LinkClock.cpp

```diff
@@ -113,6 +113,19 @@ int prLinkClock_NumPeers(struct VMGlobals* g, int numArgsPushed) {
     return errNone;
 }
 
+int prLinkClock_GetQuantum(struct VMGlobals* g, int numArgsPushed) {
+    PyrSlot* a = g->sp;
+    LinkClock* clock = (LinkClock*)slotRawPtr(&slotRawObject(a)->slots[1]);
+    if (!clock) {
+        error("clock is not running.\n");
+        return errFailed;
+    }
```

Comment:
this matches surrounding code but would prefer to see some of this factored out if you have time, the same 5 lines are repeated in every primitive here

Reply:
I don't mind to change it, but how should it be refactored? If there's a `getClockInstance()` function, then you could pass in either `g->sp` or `g->sp - 1`, easy enough, but then... if you return this, then the calling function still needs to have an `if` block to post the error and return. Or, leave the stack access as is and the factored-out function could handle the null pointer case -- but the calling function would still need an `if` block to return the error code to its caller.

Sure, code duplication is usually a red flag, but at my admittedly low C++ skill level, I don't see how factoring it out will make it more concise (and if it doesn't make it more concise, then what's the point?).

Reply:
one simple idea -

```cpp
LinkClock* tryGetLinkClock(PyrSlot* clockSlot) {
    auto* clock = static_cast<LinkClock*>(slotRawPtr(slotRawObject(clockSlot)->slots + 1));
    if (!clock) {
        error("clock is not running.\n");
    }
    return clock;
```

```cpp
int prLinkClock_GetQuantum(...) {
    if (auto* clock = tryGetLinkClock(g->sp)) {
        SetFloat(g->sp, clock->GetQuantum());
        return errNone;
    } else {
        return errFailed;
    }
}
```

---

another idea, kind of like a continuation, doesn't contain an extra branch like the idea above. if we had std::optional with a monadic interface we could use `map()`

```cpp
// Expected signature for F: `int(LinkClock*)`, returning an error code
template <typename F>
void getLinkClockAndThen(PyrSlot* clockSlot, F&& continuation) {
    auto* clock = static_cast<LinkClock*>(slotRawPtr(slotRawObject(clockSlot)->slots + 1));
    if (clock) {
        return continuation(clock);
    } else {
        error("clock is not running.\n");
        return errFailed;
    }
}
```

```cpp
int prLinkClock_GetQuantum(...) {
    getLinkClockAndThen(g->sp, [](const LinkClock* clock) {
        SetFloat(g->sp, clock->GetQuantum());
        return errNone;
    });
}
```

```cpp
int prLinkClock_SetQuantum(...) {
    getLinkClockAndThen(g->sp - 1, [g](LinkClock* clock) {
        double quantum;
        if (slotDoubleVal(g->sp, &quantum)) {
            return errWrongType;
        } else {
            clock->SetQuantum(quantum);
            return errNone;
        }
    });
}
```

---

you could even get it down to something like

```cpp
int prLinkClock_GetQuantum(...) { return tryGetLinkClockField(g->sp, &LinkClock::GetQuantum); }
int prLinkClock_SetQuantum(...) { return trySetLinkClockField(g->sp, &LinkClock::SetQuantum); }
int prLinkClock_GetLatency(...) { return tryGetLinkClockField(g->sp, &LinkClock::GetLatency); }
int prLinkClock_SetLatency(...) { return trySetLinkClockField(g->sp, &LinkClock::SetLatency); }

// at which point you may as well refactor it so you can write this:
definePrimitive(..., "_LinkClock_GetQuantum", &tryGetLinkClockField<LinkClock::GetQuantum>);
```

but that would depend on type assumptions that are a bit fragile.

happy to explain any of this further if you'd like me to. also feel free to not do any of it in this PR.

Reply:
I ended up taking the first, simplest option. I agree that it reads better.

I'm not interested enough in C++ to wrap my head around that last one :rofl: The continuation does make sense to me (since I'm accustomed to first class functions in sclang) but I lean towards a simpler C++ idiom. (One reason, for instance, why I haven't touched supernova code is that I just don't understand it, and don't have time or interest to get there.)

Reply:
> I'm not interested enough in C++ to wrap my head around that last one 🤣

tbh, i find it a little frustrating that you asked me to help you understand how to refactor this, and are now saying you're not interested in understanding it after i made an attempt to explain it.

> I lean towards a simpler C++ idiom. 

so do i, which is why i prefer the solution that uses standard library features to eliminate a large amount of duplicated code.

Reply:
Oh, now I see the "reply" field... it wasn't there before.

I apologize for causing offense -- I had read the last couple of suggestions like "well, you *could* do this but *haha* of course we won't," and my intent was to respond in a similarly lighthearted way, but I see how my wording could be read differently. (I might also suggest that PR review comments could ideally stick closer to what is required to resolve the PR, without an expectation that anything extra would be taken with the same weight. The only parts that I said I wasn't interested in were rather exotic suggestions that we would never have used.)

I incorporated the first of those suggestions -- agreed that it reads better this way.